### PR TITLE
Do not use ObjectMapper directly

### PR DESCRIPTION
**Pull request by wendigo**
It doesn't have the configuration that we need

Follow-up to https://github.com/trinodb/trino/issues/21356

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/scalar/json/JsonInputFunctions.java [hunk 1]
```diff
@@ -56,7 +59,13 @@ public final class JsonInputFunctions
     public static final String VARBINARY_UTF16_TO_JSON = "$varbinary_utf16_to_json";
     public static final String VARBINARY_UTF32_TO_JSON = "$varbinary_utf32_to_json";
 
-    private static final ObjectMapper MAPPER = new ObjectMapper();
+    private static final JsonFactory JSON_FACTORY = createJsonFactory();
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/operator/scalar/json/JsonInputFunctions.java:62]: inline?

**Review by losipiuk — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/util/JsonUtil.java [hunk 2]
```diff
@@ -107,17 +109,19 @@ public final class JsonUtil
...
+            .get()
+            .configure(ORDER_MAP_ENTRIES_BY_KEYS, false);
 
     private static final int MAX_JSON_LENGTH_IN_ERROR_MESSAGE = 10_000;
 
     // Note: JsonFactory is mutable, instances cannot be shared openly.
     public static JsonFactory createJsonFactory()
     {
-        return jsonFactoryBuilder().disable(CANONICALIZE_FIELD_NAMES).build();
+        return jsonFactoryBuilder().disable(CANONICALIZE_FIELD_NAMES).build().setCodec(OBJECT_MAPPED_UNORDERED);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/util/JsonUtil.java:124]: This didn't use OBJECT_MAPPED_UNORDERED before. Why this change?

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonUtils.java [hunk 3]
```diff
@@ -151,7 +151,7 @@ private static <T> T parseJson(JsonNode node, String jsonPointer, Class<T> javaT
 
     public static JsonFactory jsonFactory()
     {
-        return jsonFactoryBuilder().build();
+        return jsonFactoryBuilder().build().setCodec(OBJECT_MAPPER);
```
Comment (findepi) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonUtils.java:154]: Does this break encapsulation of OBJECT_MAPPER?

Comment (findepi) [reply lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonUtils.java:154]: it indeed does:

```
((ObjectMapper) jsonFactory().getCodec()).configure(ORDER_MAP_ENTRIES_BY_KEYS, false)
```

**Review by findepi — CHANGES_REQUESTED** [review]

**Comment by wendigo** [comment]
@findepi Is first commit ok? I'll drop the second one.

File: core/trino-main/src/main/java/io/trino/operator/scalar/json/JsonOutputFunctions.java [hunk 4]
```diff
@@ -60,7 +62,13 @@ public final class JsonOutputFunctions
     public static final String JSON_TO_VARBINARY_UTF16 = "$json_to_varbinary_utf16";
     public static final String JSON_TO_VARBINARY_UTF32 = "$json_to_varbinary_utf32";
 
-    private static final ObjectMapper MAPPER = new ObjectMapper();
+    private static final JsonFactory JSON_FACTORY = createJsonFactory();
+
+    static {
+        // Changes factory. Necessary for JsonParser.readValueAsTree to work.
+        new ObjectMapper(JSON_FACTORY);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/operator/scalar/json/JsonOutputFunctions.java:69]: the code here does not call readValueAsTree()
is the static block redundant?

**Comment by findepi** [comment]
in the commit message there is explanation:

> It doesn't have the configuration that we need

i think a reader would naturally ask -- what was the configuration that we need that was missing.

File: core/trino-main/src/main/java/io/trino/operator/scalar/json/JsonInputFunctions.java [hunk 5]
```diff
@@ -56,7 +59,13 @@ public final class JsonInputFunctions
     public static final String VARBINARY_UTF16_TO_JSON = "$varbinary_utf16_to_json";
     public static final String VARBINARY_UTF32_TO_JSON = "$varbinary_utf32_to_json";
 
-    private static final ObjectMapper MAPPER = new ObjectMapper();
```
[the remark this answers is not in the document]
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/operator/scalar/json/JsonInputFunctions.java:59]: It's called in a couple of places (i.e. tests) and in most of them it's harmless but in the engine code it should be banned because it operates on both user input and these instances live long.

[the remark this answers is not in the document]
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/operator/scalar/json/JsonInputFunctions.java:59]: makes sense. so how do we prevent new not harmless direct uses of ObjectMapper?
